### DISCUSSION TITLE
Building different machines & distros

### DISCUSSION THREAD

**Discussion by tirankas** [OE4T/tegra-demo-distro#88] [General]
Hello,
If I build for a specific machine/distro and then build again (on the same build env) for different machine/distro, it seems the second build is totally messed up.
For example:
. ./setup-env --machine jetson-nano-emmc --distro mender
bitbake image-prod

Then:
. ./setup-env --machine jetson-xavier-nx-emmc--distro mender
bitbake image-prod
I see the new image is built for  jetson-nano-emmc:
MACHINE              = "jetson-nano-emmc"

Diving into the setup-env and setup-env-internal, I see the machine/distro is only picked up if build/conf/local.conf doesnt exist.

In my old system before jumping into the tegra-demo, my setup supports switching machine/distro with no issue.

Is this done like this purposely? Arent we missing the benefits of having the capability using build artefacts from one machine/distro to another without the need to rebuild anything from scratch?

Thanks

**Comment by madisongh** [comment]
You can easily build for a different machine in a single build workspace by setting `MACHINE` in the environment.  For example
```
$ MACHINE=jetson-nano-emmc bitbake image-prod
```
In theory, you can do the same for `DISTRO`, provided that the other distro uses the exact same set of layers as the one you configured at setup time.  That is *not* the case, however, for `tegrademo` and `tegrademo-mender`, since the `-mender` variant adds the Mender-related layers to `conf/bblayers.conf`.

And in practice, I've found, at least, that it's unsafe to assume that building different distros (for any Yocto-based setup, not just here) in the same workspace will ever work, in the general case.  You're unlikely to get very much reuse (other than downloads) anyway, since distro-level configuration settings are likely to be different, and that affects all recipes in the build.

I usually have multiple build directories, one per distro in my setups:
```
$ . ./setup-env --machine jetson-xavier-nx-emmc --distro tegrademo
```
(which creates the default `build` subdirectory) and (in another window)
```
$ . ./setup-env --machine jetson-xavier-nx-emmc --distro tegrademo-mender build-mender
```
which creates `build-mender` for the other distro.

I then modify the `DL_DIR` and `SSTATE_DIR` settings to point to directories outside the build workspace for sharing between them.
